smithy-model/src/main/java/software/amazon/smithy/model/loader/sourcecontext/DefaultSourceLoader.java

```diff
@@ -0,0 +1,229 @@
+        } catch (IOException e) {
+            throw new UncheckedIOException("Unable to load source location context for " + source, e);
+        }
+    }
+
+    private boolean isValidLine(int line) {
+        line--;
+        return line >= 0 && line < lines.size();
+    }
+
+    private void addLineIfValid(int line, List<Line> mutate) {
+        if (isValidLine(line)) {
+            mutate.add(lines.get(line - 1));
+        }
+    }
+
+    private List<Line> parseLines(FromSourceLocation source) {
+        SourceLocation location = source.getSourceLocation();
+        int line = location.getLine();
+
+        if (source instanceof ValidationEvent) {
+            ValidationEvent event = (ValidationEvent) source;
+            Shape targetShape = event.getShapeId().flatMap(model::getShape).orElse(null);
+            if (targetShape != null) {
+                if (targetShape.getSourceLocation().equals(location)) {
+                    // The validation event is on a shape since the location is equal to the shape location.
+                    source = targetShape;
+                } else if (targetShape.getSourceLocation().getLine() > location.getLine()) {
+                    // Assume it's a trait in the Smithy IDL. Show the trait definition followed by shape definition.
+                    return parseTraitBeforeShape(location, targetShape);
+                } else if (location.getFilename().endsWith(".json")) {
+                    // Assume it's a trait defined in JSON. Show the trait definition followed by shape definition.
+                    return parseTraitAfterShape(location, targetShape);
+                } else {
+                    // It's a trait that uses "apply" or the location is invalid, so just show the trait line.
+                    return Collections.singletonList(lines.get(line - 1));
+                }
+            }
+        }
+
+        if (source instanceof MemberShape) {
+            return parseMemberShape(location, (MemberShape) source);
+        }
+
+        return parseOtherComponents(location);
+    }
+
+    private List<Line> parseTraitBeforeShape(SourceLocation location, Shape targetShape) {
+        List<Line> result = new ArrayList<>(2);
+        addLineIfValid(location.getLine(), result);
+        addLineIfValid(targetShape.getSourceLocation().getLine(), result);
+        return result;
+    }
+
+    private List<Line> parseTraitAfterShape(SourceLocation location, Shape targetShape) {
+        List<Line> result = new ArrayList<>(2);
+        addLineIfValid(targetShape.getSourceLocation().getLine(), result);
+        addLineIfValid(location.getLine(), result);
+        return result;
+    }
+
+    private List<Line> parseMemberShape(SourceLocation location, MemberShape member) {
+        // Members should always crawl up to the defining shape in both the IDL and JSON.
+        Shape container = model.getShape(member.getContainer()).orElse(null);
```

Comment:
Is there a reason you'd expect the container to not be present, or is this just in case?

Reply:
Yeah just in case. I added logging here and a comment.